pkg/abcupdater/version.go

```diff
@@ -91,45 +91,44 @@ To disable notifications for this new version, set {{.OptOutEnvVar}}="{{.Current
 	maxErrorResponseBytes = 2048
 )
 
-func (c *CheckVersionParams) loadLocalCachedData() (*LocalVersionData, error) {
-	var path string
-	if c.CacheFileOverride != "" {
-		path = c.CacheFileOverride
-	} else {
-		dir, err := localstore.DefaultDir(c.AppID)
-		if err != nil {
-			return nil, fmt.Errorf("could not calculate cache path: %w", err)
-		}
-		path = filepath.Join(dir, localVersionFileName)
-	}
-	var cached LocalVersionData
-	err := localstore.LoadJSONFile(path, &cached)
-	if err != nil {
-		return nil, fmt.Errorf("could not load cached data: %w", err)
+// CheckAppVersion calls CheckAppVersionSync in a go routine. It returns a closure
+// to be run after program logic which will block until a response is returned
+// or provided context is canceled. If no provided deadline in context, defaults to 2 seconds.
+// If there is an update, out() will be called during the
+// returned closure.
+//
+// If no update is available: out() will not be called.
+// If there is an error: out() will not be called, message will be logged as WARN.
+// If the context is canceled: out() is not called.
+// If processing config fails: an error will be returned synchronously.
+// Example out(): `func(s string) {fmt.Fprintln(os.Stderr, s)}`.
+func CheckAppVersion(ctx context.Context, params *CheckVersionParams, out func(string)) (func(), error) {
+	cancel := func() {}
+	if _, ok := ctx.Deadline(); !ok {
+		ctx, cancel = context.WithTimeout(ctx, time.Second*2) //nolint:govet
 	}
-	return &cached, nil
-}
 
-func (c *CheckVersionParams) setLocalCachedData(data *LocalVersionData) error {
-	var path string
-	if c.CacheFileOverride != "" {
-		path = c.CacheFileOverride
-	} else {
-		dir, err := localstore.DefaultDir(c.AppID)
-		if err != nil {
-			return fmt.Errorf("could not calculate cache path: %w", err)
-		}
-		path = filepath.Join(dir, localVersionFileName)
+	lookuper := params.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
 	}
-	if err := localstore.StoreJSONFile(path, data); err != nil {
-		return fmt.Errorf("could not cache version: %w", err)
+	var c config
+	if err := envconfig.ProcessWith(ctx, &envconfig.Config{
+		Target:   &c,
+		Lookuper: lookuper,
+	}); err != nil {
+		// This leaks context. OK since only runs once, and timeout is short.
```

Comment:
I'm not quite sure what this means. I think maybe it means that the error handler returns from the function without calling cancel(), and therefore leaks some resources associated with the deadline handling. If I'm right about that, can we just call cancel() here, rather than documenting that it's not called? Maybe I'm missing something.

Reply:
> can we just call cancel() here,

That is so painfully obvious now that you mention it